browser/src/layer/tile/CanvasTileLayer.js

```diff
@@ -5399,26 +5413,74 @@ L.CanvasTileLayer = L.Layer.extend({
 			var centerTwips = this._pixelsToTwips(centerPos);
 
 			// Perform action
-			console.log('_debugMoveSheetTimeout move to ',pos,' click at ', centerPos, centerTwips);
+			console.log('Automated User: Move to ',pos,' click at ', centerPos, centerTwips);
 			this._map.fire('scrollto', pos);
 			this._postMouseEvent('buttondown', centerTwips.x, centerTwips.y, 1, 1, 0);
 			this._postMouseEvent('buttonup', centerTwips.x, centerTwips.y, 1, 1, 0);
 		}
 	},
 
-	_debugAddRemoveShapesTimeout: function () {
-		var shapes = ['smiley','sun','moon','lightning','heart','flower'];
-		var shape = shapes[Math.floor(Math.random() * shapes.length)];
-		this._map.sendUnoCommand('.uno:SymbolShapes.'+shape);
-
-		setTimeout(L.bind(function() {app.socket.sendMessage('removetextcontext id=0 before=0 after=1');},this), 2000);
+	_debugAutomatedUserInputTimeout: function () {
+		switch (this._debugAutomatedUserInputCase) {
+			case 0:
```

Comment:
This looks fine; I'd prefer to have a set of tools that are added into this state machine and each of which has a mini-state-machine of its own - and the outer tool wrapper manages those to make sure the overall number range 1->100 is split into ten 0 -> 10 state machines (or for however many states they ask for). This infrastructure would then allow tools to be included into the larger state-machine loop at will with toggles =) clearly on registration they'd need to state how many states they want to request; and perhaps the time for each one - up to you =)

We should then do typing with a reasonable cadence - not throwing lots of keys in at a time which is unlike a real user - but typing one every 100ms or so.

Otherwise this looks promising =)